Client/Frontend/Browser/BrowserViewController/BrowserViewController+WKNavigationDelegate.swift

```diff
@@ -529,45 +529,87 @@ extension BrowserViewController: WKNavigationDelegate {
+  @MainActor
+  public func webView(_ webView: WKWebView, respondTo challenge: URLAuthenticationChallenge) async -> (URLSession.AuthChallengeDisposition, URLCredential?) {
 
     // If this is a certificate challenge, see if the certificate has previously been
     // accepted by the user.
     let origin = "\(challenge.protectionSpace.host):\(challenge.protectionSpace.port)"
     if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
       let trust = challenge.protectionSpace.serverTrust,
       let cert = SecTrustGetCertificateAtIndex(trust, 0), profile.certStore.containsCertificate(cert, forOrigin: origin) {
-      completionHandler(.useCredential, URLCredential(trust: trust))
-      return
+      return (.useCredential, URLCredential(trust: trust))
+    }
+    
+    // Certificate Pinning
+    if challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust {
+      if let serverTrust = challenge.protectionSpace.serverTrust {
+        let host = challenge.protectionSpace.host
+        let port = challenge.protectionSpace.port
+        
+        return await Task.detached(priority: .userInitiated) {
+          let result = BraveCertificateUtility.verifyTrust(serverTrust,
+                                                           host: host,
+                                                           port: port)
+          let certificateChain = (0..<SecTrustGetCertificateCount(serverTrust))
+            .compactMap { SecTrustGetCertificateAtIndex(serverTrust, $0) }
+          
+          // Cert is valid and should be pinned
+          if result == 0 {
+            return (.useCredential, URLCredential(trust: serverTrust))
+          }
+          
+          // Cert is valid and should not be pinned
+          // Let the system handle it and we'll show an error if the system cannot validate it
+          if result == Int32.min {
+            return (.performDefaultHandling, nil)
+          }
+          
+          return await Task { @MainActor in
+            // Cert is invalid and cannot be pinned
+            Logger.module.error("CERTIFICATE_INVALID")
+            let errorCode = CFNetworkErrors.braveCertificatePinningFailed.rawValue
+            
+            let underlyingError = NSError(domain: kCFErrorDomainCFNetwork as String,
+                                          code: Int(errorCode),
+                                          userInfo: ["_kCFStreamErrorCodeKey": Int(errorCode)])
+            
+            let error = NSError(domain: kCFErrorDomainCFNetwork as String,
+                                code: Int(errorCode),
+                                userInfo: [NSURLErrorFailingURLErrorKey: webView.url as Any,
+                                           "NSErrorPeerCertificateChainKey": certificateChain,
```

Comment:
Private API key?

Reply:
Yeah. It's the only way to get the certificate chain from `NSError`. Fortunately, I'm the one setting it here, so even if Apple removes this key, it won't break anything as I'm the one setting it, and reading it back elsewhere in the code (certificate viewer).